crates/runtime/src/vmcontext.rs

```diff
@@ -434,6 +436,62 @@ impl VMGlobalDefinition {
         Self { storage: [0; 16] }
     }
 
+    /// Create a `VMGlobalDefinition` from a `ValRaw`.
+    ///
+    /// # Unsafety
+    ///
+    /// This raw value's type must match the given `WasmValType`.
+    pub unsafe fn from_val_raw(wasm_ty: WasmValType, raw: ValRaw) -> Self {
+        let mut global = Self::new();
+        match wasm_ty {
+            WasmValType::I32 => *global.as_i32_mut() = raw.get_i32(),
+            WasmValType::I64 => *global.as_i64_mut() = raw.get_i64(),
+            WasmValType::F32 => *global.as_f32_bits_mut() = raw.get_f32(),
+            WasmValType::F64 => *global.as_f64_bits_mut() = raw.get_f64(),
+            WasmValType::V128 => *global.as_u128_bits_mut() = raw.get_v128().to_ne_bytes(),
+            WasmValType::Ref(r) => match r.heap_type {
+                WasmHeapType::Extern => {
+                    global.init_gc_ref(VMGcRef::from_raw_u32(raw.get_externref()))
+                }
+                WasmHeapType::Any | WasmHeapType::I31 | WasmHeapType::None => {
+                    global.init_gc_ref(VMGcRef::from_raw_u32(raw.get_anyref()))
+                }
+                WasmHeapType::Func | WasmHeapType::Concrete(_) | WasmHeapType::NoFunc => {
+                    *global.as_func_ref_mut() = raw.get_funcref().cast()
+                }
+            },
+        }
+        global
+    }
+
+    /// Get this global's value as a `ValRaw`.
+    ///
+    /// # Unsafety
+    ///
+    /// This global's value's type must match the given `WasmValType`.
+    pub unsafe fn to_val_raw(&self, gc_store: &mut GcStore, wasm_ty: WasmValType) -> ValRaw {
+        match wasm_ty {
+            WasmValType::I32 => ValRaw::i32(*self.as_i32()),
+            WasmValType::I64 => ValRaw::i64(*self.as_i64()),
+            WasmValType::F32 => ValRaw::f32(*self.as_f32_bits()),
+            WasmValType::F64 => ValRaw::f64(*self.as_f64_bits()),
+            WasmValType::V128 => ValRaw::v128(*self.as_u128()),
+            WasmValType::Ref(r) => match r.heap_type {
+                WasmHeapType::Extern => ValRaw::externref(
+                    self.as_gc_ref()
+                        .map_or(0, |r| gc_store.clone_gc_ref(r).as_raw_u32()),
+                ),
+                WasmHeapType::Any | WasmHeapType::I31 | WasmHeapType::None => ValRaw::anyref(
+                    self.as_gc_ref()
+                        .map_or(0, |r| gc_store.clone_gc_ref(r).as_raw_u32()),
+                ),
+                WasmHeapType::Func | WasmHeapType::Concrete(_) | WasmHeapType::NoFunc => {
```

Comment:
Bit of a side comment, but I get a little uncomfortable with `Concrete(_)` here since it seems very easy to overlook this in the future when the concrete type may be a function or may be a struct. Is it possible to perhaps restructure this or add an assert? 

It seems especially relevant here and above given how unsafe the methods are

Reply:
Unfortunately there isn't a good way to resolve this, given the `wasmtime` vs `wasmtime-runtime` crate split. We would need to take a `wasmtime::Engine` here and then look up the type index in the engine's types registry and match on the resulting value (which can only be a function type right now). But we can't get an engine inside the `wasmtime-runtime` crate.

Reply:
We have to solve this in the near future though right?

Originally I was hoping we could have `Concrete` renamed to `ConcreteFunc` and then also have `Concrete{Struct,Array}` or something like that, but I'm not sure if that's still possible.

Do you have other ideas though on what to do here once arrays/structs are implemented?

Reply:
My plan was to start passing in a function to these sorts of methods that need to do the index-to-type resolution. Between that, fuzzing, and using LSP to find all uses of `Concrete(_)` I wasn't too too worried about missing a spot to update.

The other thing we could do is try moving the type registry into `wasmtime-runtime`.

Reply:
I may be blowing things out of proportion here perhaps, but I'm personally pretty worried about this. I just ran across [another instance](https://github.com/bytecodealliance/wasmtime/blob/1f8c3df2d305c4e8e32529911bd97fe376f90bd7/crates/wasmtime/src/runtime/types.rs#L592-L597) of where type information will be needed as well.

I'm basically pretty worried that we've pretty deeply baked in the assumption that "`Concrete` is always a function" and I'd like to avoid baking it in any further, especially here where it's necessary for safety. 

I suppose for the sake of this PR could you add an assertion to the constructors of `Concrete` that it's always a function? (or double-check such an assertion already exists?) More-or-less we need to re-audit all usages of `Concrete` when the GC implementation gets further along.

Reply:
(We talked offline about this and decided that renaming `Concrete(_)` to `ConcreteFunc(_)` actually is okay, and will do it in a future PR. It wasn't workable for `wasmparser`, since that crate needs to be able to parse opcodes without having a validated types section where we can look things up in and determine whether we are talking about a concrete func or e.g. struct, but that constraint doesn't apply here.)